aptos-move/move-examples/post_mint_reveal_nft/sources/minting.move

```diff
@@ -0,0 +1,1048 @@
+        let collection_config = borrow_global_mut<CollectionConfig>(@post_mint_reveal_nft);
+
+        assert!(vector::length(&token_uris) + big_vector::length(&collection_config.tokens) <= collection_config.collection_maximum || collection_config.collection_maximum == 0, error::invalid_argument(EEXCEEDS_COLLECTION_MAXIMUM));
+
+        let i = 0;
+        while (i < vector::length(&token_uris)) {
+            big_vector::push_back(&mut collection_config.tokens, TokenAsset {
+                token_uri: *vector::borrow(&token_uris, i),
+                property_keys: *vector::borrow(&property_keys, i),
+                property_values: *vector::borrow(&property_values, i),
+                property_types: *vector::borrow(&property_types, i),
+            });
+            i = i + 1;
+        };
+    }
+
+    /// Mint source certificate.
+    public entry fun mint_source_certificate(
+        nft_claimer: &signer,
+        amount: u64
+    ) acquires NFTMintConfig, PublicMintConfig, WhitelistMintConfig, SourceToken, CollectionConfig {
+        assert!(exists<CollectionConfig>(@post_mint_reveal_nft) && exists<WhitelistMintConfig>(@post_mint_reveal_nft) && exists<PublicMintConfig>(@post_mint_reveal_nft), error::permission_denied(ECONFIG_NOT_INITIALIZED));
+
+        let whitelist_mint_config = borrow_global_mut<WhitelistMintConfig>(@post_mint_reveal_nft);
+        let public_mint_config = borrow_global_mut<PublicMintConfig>(@post_mint_reveal_nft);
+        let collection_config = borrow_global_mut<CollectionConfig>(@post_mint_reveal_nft);
+
+        let now = timestamp::now_seconds();
+        let is_whitelist_minting_time = now > whitelist_mint_config.whitelist_minting_start_time && now < whitelist_mint_config.whitelist_minting_end_time;
+        let is_public_minting_time = now > public_mint_config.public_minting_start_time && now < public_mint_config.public_minting_end_time;
+        assert!(is_whitelist_minting_time || is_public_minting_time, error::permission_denied(EMINTING_IS_NOT_ENABLED));
+        let claimer_addr = signer::address_of(nft_claimer);
+
+        // if this is the whitelist minting time
+        if (is_whitelist_minting_time) {
+            assert!(bucket_table::contains(&whitelist_mint_config.whitelisted_address, &claimer_addr), error::permission_denied(EACCOUNT_NOT_WHITELISTED));
+            let remaining_mint_allowed = bucket_table::borrow_mut(&mut whitelist_mint_config.whitelisted_address, claimer_addr);
+            assert!(*remaining_mint_allowed >= amount, error::invalid_argument(EAMOUNT_EXCEEDS_MINTS_ALLOWED));
+            while (amount > 0) {
+                mint_source_certificate_internal(nft_claimer, whitelist_mint_config.whitelist_mint_price);
+                amount = amount - 1;
+                *remaining_mint_allowed = *remaining_mint_allowed - 1;
+            };
```

Comment:
This should be a batch api. Otherwise the gas is multiplied by a loop iterations, especially when `mint_token` takes an `amount` as a param.